ENH: refactor LaplacianSharpeningImageFilter

ntustison opened the pull request:
PR based on https://github.com/InsightSoftwareConsortium/ITK/issues/4051.

This commit fixes the non-use of the ``m_UseImageSpacing`` variable.   The filter also needed to be refactored given some of the streaming incompatibilities.  I believe what happened was that the original author also wrote the ``itkLaplacianImageFilter``, which is streaming-capable, and copied much of the code.  Part of the refactoring included replacing much of the borrowed code with a call to the actual Laplacian filter.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

blowekamp commented on Modules/Filtering/ImageFeature/include/itkLaplacianSharpeningImageFilter.hxx:
```diff
@@ -35,136 +33,53 @@ LaplacianSharpeningImageFilter<TInputImage, TOutputImage>::PrintSelf(std::ostrea
-    throw e;
-  }
-}
-
```
The following is the preferred pattern to when a filter must produce all of it's outputs and requires all of it's inputs:
```
EnlargeOutputRequestedRegion(
    DataObject * output)
  {
    Superclass::EnlargeOutputRequestedRegion(output);

    if (output != nullptr)
    {
      output->SetRequestedRegionToLargestPossibleRegion();
    }
  }

```

blowekamp commented on Modules/Filtering/ImageFeature/include/itkLaplacianSharpeningImageFilter.hxx:
```diff
@@ -35,136 +33,53 @@ LaplacianSharpeningImageFilter<TInputImage, TOutputImage>::PrintSelf(std::ostrea
-  filter->GetOutput()->SetRequestedRegion(this->GetOutput()->GetRequestedRegion());
+  using LaplacianImageFilter = LaplacianImageFilter<InputImageType, RealImageType>;
+  typename LaplacianImageFilter::Pointer laplacianFilter = LaplacianImageFilter::New();
+  laplacianFilter->SetInput(this->GetInput());
```
The internal mini-pipeline needs to be isolated from the external. This is done with graft the input:
```
auto localInput = TInputImage::New();
localInput->Graft(this->GetInput());
laplacianFilter->SetInput(localInput);
```

ntustison replied:
Thanks @blowekamp .  

Excuse the ignorance but I'm having a hard time following you.  First, is this assuming that this filter is capable of  streaming?  Because I assumed it wasn't and that all the "streaming" remnants were an artifact of copying over from the LaplacianImageFilter, which is capable of streaming.

ntustison replied:
Ah, okay.  I'm a little slow this morning but I understand now (at least this part).  Thanks.

blowekamp commented on Modules/Filtering/ImageFeature/include/itkLaplacianSharpeningImageFilter.hxx:
```diff
@@ -35,136 +33,53 @@ LaplacianSharpeningImageFilter<TInputImage, TOutputImage>::PrintSelf(std::ostrea
   typename MinimumMaximumImageCalculator<InputImageType>::Pointer inputCalculator =
     MinimumMaximumImageCalculator<InputImageType>::New();
   typename MinimumMaximumImageCalculator<RealImageType>::Pointer filteredCalculator =
     MinimumMaximumImageCalculator<RealImageType>::New();
```
The MinimumMaximumImageFilter would be multithreaded.

blowekamp commented on Modules/Filtering/ImageFeature/include/itkLaplacianSharpeningImageFilter.hxx:
```diff
@@ -183,51 +98,41 @@ LaplacianSharpeningImageFilter<TInputImage, TOutputImage>::GenerateData()
   {
-    value = it.Get();
-
     // adjust value to make the mean intensities before and after match
```
The UnaryGeneratorImageFilter can except lambda functions which should make converting the code to run multi-threaded in a filter relatively easy.

blowekamp commented:
@ntustison Thanks for tackling this! Hope I didn't give too many suggestions. The only items which I would consider bugs:
 - The missing `EnlargeOutputRequestedRegion` method
 - The missing `Graft`-ed image to isolate the pipeline.
Thanks! 🤩

ntustison commented:
Okay @blowekamp , I was finally able to get a clean set of checks.  A couple notes:

As you asked for, I believe I addressed

* The missing EnlargeOutputRequestedRegion method
* The missing Graft-ed image to isolate the pipeline.

I also removed all the iterators, addressing something that you alluded to in your initial review.  I also more closely coordinated with the related ``itkLaplacianImageFilter`` by adding the same concept checking and changing the examples and wrapping accordingly.  Let me know of anything else when you get a chance.

blowekamp commented:
Wow, nice work. The changes look really good. 👍 

Does the filter now only work with float images?

ntustison commented:
> Does the filter now only work with float images?

Currently, yes.  But this decision was based simply on harmonizing with the ``LaplacianImageFilter`` which also only works with real images.  Since the first operation for Laplacian sharpening is invoking the Laplacian filter, it would seem to me that they should be consistent as to the allowed input image type.  Whether or not we want to limit both to only working with real images (as it is with this proposed pull request) or expand both to working with all scalar images, I don't have a strong opinion and defer to people, such as yourself, with more knowledge of the intricacies of the toolkit.

blowekamp commented on Modules/Filtering/ImageFeature/include/itkLaplacianSharpeningImageFilter.hxx:
```diff
@@ -18,217 +18,130 @@
-  }
+  // Shift and window the output
+
+  using IntensityWindowingFilterType = IntensityWindowingImageFilter<RealImageType, OutputImageType>;
```
Haven't we already done the intensity shift with the arithmetic operations? Do we need to clamp?

ntustison replied:
I didn't scrutinize the logic of the code to see if the clamping is unnecessary here.  However, as the original iterator did check and perform clamping, if necessary, I just recreated it with the windowing filter.

blowekamp replied:
I'm still now sure the math matches. Do the test have a base line image with values to ensure that there is not much difference in output after this refactoring?

ntustison replied:
Exactly.  I wasn't sure, either, and was also concerned given the wholesale replacement of sections with the various filters.  Fortunately there is a baseline test with an image difference check and it does appear to pass on all the platforms.

dzenanz reviewed (commented):
Is this ready for merging? I guess it should be squashed, before or upon merge.

ntustison commented:
Hey @blowekamp, just wanted to follow up and see if there was anything lacking with this pull request so I can check it off my to-do list.    Thanks.

dzenanz commented:
Maybe @blowekamp is on vacation? Or was? 😄

blowekamp commented:
Sorry for the delayed response.

If everyone is comfortable that the level of testing ensure the same results, I think it's good. This is quite a nice improvement to the filter.

After it's merged I may run it's passes through in SimpleITK and give the filter an additional close look.

dzenanz commented:
I used GitHub's "Squash and Merge" feature.

ntustison commented:
Thanks @blowekamp for all the help.  And thanks @dzenanz and @hjmjohnson for pointing to the Squash and Merge feature.  That's a good one to remember.
